cpp/src/fil/treelite_import.cu

```diff
@@ -691,48 +594,66 @@ struct tl2fil_sparse_check_t<sparse_node8> {
   }
 };
 
-// uses treelite model with additional tl_params to initialize FIL params,
-// trees (stored in *ptrees) and sparse nodes (stored in *pnodes)
 template <typename fil_node_t, typename threshold_t, typename leaf_t>
-void tl2fil_sparse(std::vector<int>* ptrees,
-                   std::vector<fil_node_t>* pnodes,
-                   forest_params_t* params,
-                   const tl::ModelImpl<threshold_t, leaf_t>& model,
-                   const treelite_params_t* tl_params,
-                   cat_sets_owner* cat_sets,
-                   std::vector<float>* vector_leaf)
-{
-  tl2fil_common(params, model, tl_params);
-  tl2fil_sparse_check_t<fil_node_t>::check(model);
+struct tl2fil_t {
+  std::vector<int> trees;
+  std::vector<fil_node_t> nodes;
+  std::vector<float> vector_leaf;
+  forest_params_t params;
+  cat_sets_owner cat_sets;
+  const tl::ModelImpl<threshold_t, leaf_t>& model;
+  const treelite_params_t& tl_params;
 
-  size_t num_trees = model.trees.size();
+  tl2fil_t(const tl::ModelImpl<threshold_t, leaf_t>& model_, const treelite_params_t& tl_params_)
```

Comment:
It's better to limit the constructor to basic initialization only, and perform other checks, allocations and actions in a separate method, e.g. `init()`.